src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameParser.cs

```diff
@@ -0,0 +1,455 @@
+            if (text.IsDefault)
+            {
+                return null;
+            }
+
+            try
+            {
+                return new StackFrameParser(text).TryParseTree();
+            }
+            catch (InsufficientExecutionStackException)
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Constructs a <see cref="VirtualCharSequence"/> and calls <see cref="TryParse(VirtualCharSequence)"/>
+        /// </summary>
+        public static StackFrameTree? TryParse(string text)
+            => TryParse(VirtualCharSequence.Create(0, text));
+
+        /// <summary>
+        /// Attempts to parse the full tree. Returns null on malformed data
+        /// </summary>
+        private StackFrameTree? TryParseTree()
+        {
+            var (_, methodDeclaration) = TryParseMethodDeclaration();
+            if (methodDeclaration is null)
+            {
+                return null;
+            }
+
+            var fileInformationResult = TryParseFileInformation();
+            if (!fileInformationResult.Success)
+            {
+                return null;
+            }
+
+            var remainingTrivia = _lexer.TryScanRemainingTrivia();
+
+            var eolToken = CurrentCharAsToken().With(leadingTrivia: remainingTrivia.HasValue ? ImmutableArray.Create(remainingTrivia.Value) : ImmutableArray<StackFrameTrivia>.Empty);
+
+            Debug.Assert(_lexer.Position == _lexer.Text.Length);
+            Debug.Assert(eolToken.Kind == StackFrameKind.EndOfLine);
+
+            var root = new StackFrameCompilationUnit(methodDeclaration, fileInformationResult.Value, eolToken);
+
+            return new(_lexer.Text, root);
+        }
+
+        /// <summary>
+        /// Attempts to parse the full method declaration, optionally adding leading whitespace as trivia. Includes
+        /// all of the generic indicators for types, 
+        /// 
+        /// Ex: [|MyClass.MyMethod(string s)|]
+        /// </summary>
+        private ParseResult<StackFrameMethodDeclarationNode> TryParseMethodDeclaration()
+        {
+            var (success, identifierNode) = TryParseNameNode(scanAtTrivia: true);
+            if (!success)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            if (identifierNode is not StackFrameQualifiedNameNode memberAccessExpression)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            var typeArgumentsResult = TryParseTypeArguments();
+            if (!typeArgumentsResult.Success)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
+
+            var argumentsResult = ParseMethodParameters();
+            if (!argumentsResult.Success || argumentsResult.Value is null)
+            {
+                return ParseResult<StackFrameMethodDeclarationNode>.Abort;
+            }
```

Comment:
so a core distinction here are things that are optional versus things that are not.

non-optional things don't need ParseResult.  They can just return a nullable value, where null indicates: abort.
optional things do need parse result.  because they need to be able to distinguish: "failure happened and we need to stop", vs "everything is fine, i just had no value to give you".

Reply:
> vs "everything is fine, I just had no value to give you" 

From a purely architectural point of view, I tried to make it so methods don't know if the caller requires anything from them. They know cases where they return nothing but haven't gotten into a bad state (success with null). I was wavering on naming because then everything could just be `ParseXXX` instead of `TryParseXXX`. In general I like this design because it keeps expectations the same across the file. Method definitions know they _need_ parameters, but the parameter parsing code doesn't need to know that.

Reply:
Talked offline